src/Php/ComposerPhpVersionFactory.php

```diff
@@ -0,0 +1,120 @@
+<?php declare(strict_types = 1);
+
+namespace PHPStan\Php;
+
+use Composer\Semver\VersionParser;
+use Nette\Utils\Json;
+use Nette\Utils\JsonException;
+use Nette\Utils\Strings;
+use PHPStan\File\CouldNotReadFileException;
+use PHPStan\File\FileReader;
+use PHPStan\ShouldNotHappenException;
+use function count;
+use function end;
+use function is_array;
+use function is_file;
+use function is_int;
+use function is_string;
+use function sprintf;
+
+final class ComposerPhpVersionFactory
+{
+
+	private ?PhpVersion $minVersion = null;
+
+	private ?PhpVersion $maxVersion = null;
+
+	/**
+	 * @param string[] $composerAutoloaderProjectPaths
+	 * @param int|array{min: int, max: int}|null $phpVersion
+	 */
+	public function __construct(
+		private array $composerAutoloaderProjectPaths,
+		int|array|null $phpVersion,
+	)
+	{
+		if (is_int($phpVersion)) {
+			return;
+		}
+
+		if (is_array($phpVersion)) {
+			if ($phpVersion['max'] < $phpVersion['min']) {
+				throw new ShouldNotHappenException('Invalid PHP version range: phpVersion.max should be greater or equal to phpVersion.min.');
+			}
+
+			$this->minVersion = new PhpVersion($phpVersion['min']);
+			$this->maxVersion = new PhpVersion($phpVersion['max']);
+
+			return;
+		}
+
+		// fallback to composer.json based php-version constraint
+		$composerPhpVersion = $this->getComposerRequireVersion();
+		if ($composerPhpVersion === null) {
+			return;
+		}
+
+		$parser = new VersionParser();
+		$constraint = $parser->parseConstraints($composerPhpVersion);
+
+		if (!$constraint->getLowerBound()->isZero()) {
+			$minVersion = $this->buildVersion($constraint->getLowerBound()->getVersion());
+
+			if ($minVersion !== null) {
+				$this->minVersion = new PhpVersion($minVersion->getVersionId());
+			}
+		}
+		if ($constraint->getUpperBound()->isPositiveInfinity()) {
+			return;
+		}
+
+		$this->maxVersion = $this->buildVersion($constraint->getUpperBound()->getVersion());
```

Comment:
Don't do this work in the construtor, but lazily in the getters. (Of course it should still be cached in the properties.)